cpp/src/svm/sparse_util.cuh

```diff
@@ -0,0 +1,253 @@
+
+  int row_idx = row_indices[idx];
+
+  int in_offset  = indptr_in[row_idx];
+  int row_length = indptr_in[row_idx + 1] - in_offset;
+  int out_offset = indptr_out[idx];
+  for (int pos = threadIdx.x; pos < row_length; pos += blockDim.x) {
+    indices_out[out_offset + pos] = indices_in[in_offset + pos];
+    data_out[out_offset + pos]    = data_in[in_offset + pos];
+  }
+}
+
+template <typename math_t>
+void matrixRowNorm(const raft::handle_t& handle,
+                   const MLCommon::Matrix::Matrix<math_t>& matrix,
+                   math_t* target,
+                   raft::linalg::NormType norm)
+{
+  if (matrix.is_dense()) {
+    int minor = matrix.as_dense()->is_row_major() ? matrix.get_n_cols() : matrix.get_n_rows();
+    ASSERT(matrix.as_dense()->get_ld() == minor,
+           "Dense matrix rowNorm only support contiguous data");
+    raft::linalg::rowNorm(target,
+                          matrix.as_dense()->get_data(),
+                          matrix.get_n_cols(),
+                          matrix.get_n_rows(),
+                          norm,
+                          matrix.as_dense()->is_row_major(),
+                          handle.get_stream());
+  } else {
+    auto csr_matrix = matrix.as_csr();
+    raft::sparse::linalg::rowNormCsr(handle,
+                                     csr_matrix->get_indptr(),
+                                     csr_matrix->get_data(),
+                                     csr_matrix->get_nnz(),
+                                     matrix.get_n_rows(),
+                                     target,
+                                     norm);
+  }
+}
+
+template <typename math_t>
+static void copySparseRowsToDense(const int* indptr,
+                                  const int* indices,
+                                  const math_t* data,
+                                  int n_rows,
+                                  int n_cols,
+                                  math_t* output,
+                                  const int* row_indices,
+                                  int num_indices,
+                                  cudaStream_t stream)
+{
+  thrust::device_ptr<math_t> output_ptr(output);
+  thrust::fill(
+    thrust::cuda::par.on(stream), output_ptr, output_ptr + num_indices * n_cols, (math_t)0);
+
+  // copy with 1 warp per row for now, blocksize 256
+  const dim3 bs(32, 8, 1);
+  const dim3 gs(raft::ceildiv(num_indices, (int)bs.y), 1, 1);
+  extractDenseRowsFromCSR<math_t>
+    <<<gs, bs, 0, stream>>>(output, indptr, indices, data, row_indices, num_indices);
+  cudaDeviceSynchronize();
+  RAFT_CUDA_TRY(cudaPeekAtLastError());
+}
+
+struct rowsize : public thrust::unary_function<int, int> {
+  const int* indptr_;
+  rowsize(const int* indptr) : indptr_(indptr) {}
+
+  __device__ int64_t operator()(const int& x) const { return indptr_[x + 1] - indptr_[x]; }
+};
+
+template <typename math_t>
+static void extractRows(const MLCommon::Matrix::Matrix<math_t>& matrix_in,
```

Comment:
I think this should ultimately be moved over to `raft::matrix::gather` and `raft::sparse::matrix::gather`.

Reply:
Created the RAFT issue: https://github.com/rapidsai/raft/issues/1524. We should reference this in the TODO in `extractRows`.